### ISSUE TITLE
Geb error with higher-order functions

### ISSUE THREAD

**Issue by lukaszcz** [anoma/geb#94]
Trying to compile:
```
(defpackage #:ho2
  (:shadowing-import-from :geb.lambda.spec #:func #:pair)
  (:shadowing-import-from :geb.spec #:case)
  (:use #:common-lisp #:geb.lambda.spec #:geb))

(in-package :ho2)

(defparameter *entry*
  (typed
    (app
      (!->
        (!->
          (coprod
            so1
            so1)
          (coprod
            so1
            so1))
        (coprod
          so1
          so1))
      (coprod
        so1
        so1)
      (lamb
        (!->
          (!->
            (coprod
              so1
              so1)
            (coprod
              so1
              so1))
          (coprod
            so1
            so1))
        (coprod
          so1
          so1)
        (app
          (!->
            (coprod
              so1
              so1)
            (coprod
              so1
              so1))
          (coprod
            so1
            so1)
          (index 0)
          (lamb
            (coprod
              so1
              so1)
            (coprod
              so1
              so1)
            (index 0))))
      (lamb
        (!->
          (coprod
            so1
            so1)
          (coprod
            so1
            so1))
        (coprod
          so1
          so1)
        (left
          so1
          so1
          unit)))
    (coprod
      so1
      so1)))
```
I get:
```

debugger invoked on a SIMPLE-ERROR in thread
#<THREAD "main thread" RUNNING {7006560003}>:
  Types should match for application: (× (× (+ s-1 s-1) (+ s-1 s-1))
                                         (+ s-1 s-1)
                                         (+ s-1 s-1)) (+ s-1 s-1)

Type HELP for debugger help, or (SB-EXT:EXIT) to exit from SBCL.

restarts (invokable by number or by possibly-abbreviated name):
  0: [CONTINUE] Retry assertion.
  1: [ABORT   ] Exit from the current thread.

(SB-KERNEL:ASSERT-ERROR (GEB.MIXINS:OBJ-EQUALP GEB.MIXINS:DOM TYPE) 2 GEB.MIXINS:DOM (× (× #1=(+ #2=s-1 #2#) #1#) #1# #1#) TYPE (+ #1=s-1 #1#) NIL "Types should match for application: ~A ~A" (× (× #1=(+ #2=s-1 #2#) #1#) #1# #1#) (+ #1=s-1 #1#))
```

Geb version: 0.2.0

**Comment by rokopt** [comment]
For reference, here is the term in current syntax:

```
TRANS> (well-defp (list (coprod so1 so1) (fun-type (fun-type (coprod so1 so1) (coprod so1 so1)) (coprod so1 so1))) (app (lamb
      (fun-type
       (fun-type
        (coprod so1 so1)
        (coprod so1 so1))
       (coprod so1 so1))
      (app
       (index 0)
       (lamb (coprod so1 so1)
             (index 1))))
      (lamb
       (fun-type (coprod so1 so1)
                 (coprod so1 so1))
       (left so1 (unit)))))
T
```

**Comment by agureev** [comment]
This compiles with the new #97

**Comment by rokopt** [comment]
Can this be closed as fixed by #97 ?

**Comment by agureev** [comment]
This is fixed by #127
